tools/generate_release_notes.py

```diff
@@ -1,246 +1,331 @@
+                    pull.html_url,
+                )
+        all_pull_requests.update(commit_pull_requests)
+    return all_pull_requests
+
+
+def contributors(
+    commits: Iterable[Commit], pull_requests
+) -> tuple[set[NamedUser], set[NamedUser]]:
+    """Fetch code authors and reviewers.
+
+    `authors` are first authors of commits; co-authors are not included (yet).
+    `reviewers` are users, who added reviews to a merged pull request or create the
+    merge commit for one.
+    """
+    authors = set()
+    reviewers = set()
 
-parser = argparse.ArgumentParser(usage=__doc__)
-parser.add_argument("from_commit", help="The starting tag.")
-parser.add_argument("to_commit", help="The head branch.")
-parser.add_argument(
-    "--version", help="Version you're about to release.", default="0.15.0"
-)
+    for commit in commits:
+        # TODO include co-authors?
+        if commit.author:
+            authors.add(commit.author)
+        if commit.committer:
+            reviewers.add(commit.committer)
 
-args = parser.parse_args()
+    for pull in pull_requests:
+        for review in pull.get_reviews():
+            if review.user:
+                reviewers.add(review.user)
 
-for tag in repository.get_tags():
-    if tag.name == args.from_commit:
-        previous_tag = tag
-        break
-else:
-    raise RuntimeError(f"Desired tag ({args.from_commit}) not found")
+    return authors, reviewers
 
-# For some reason, go get the github commit from the commit to get
-# the correct date
-github_commit = previous_tag.commit.commit
-previous_tag_date = datetime.strptime(
-    github_commit.last_modified, "%a, %d %b %Y %H:%M:%S %Z"
-)
 
+@dataclass(frozen=True, kw_only=True)
+class MdFormatter:
+    """Format release notes in Markdown from PRs, authors and reviewers."""
 
-all_commits = list(
-    tqdm(
-        repository.get_commits(sha=args.to_commit, since=previous_tag_date),
-        desc=f"Getting all commits between {args.from_commit} " f"and {args.to_commit}",
-    )
-)
-all_hashes = {c.sha for c in all_commits}
+    pull_requests: set[PullRequest]
+    authors: set[NamedUser]
+    reviewers: set[NamedUser]
 
-authors = set()
-reviewers = set()
-committers = set()
-users = dict()  # keep track of known usernames
+    version: str = "x.y.z"
+    title_template: str = "scikit-image {version} release notes"
+    intro_template: str = """
+We're happy to announce the release of scikit-image {version}!
+scikit-image is an image processing toolbox for SciPy that includes algorithms
+for segmentation, geometric transformations, color space manipulation,
+analysis, filtering, morphology, feature detection, and more.
 
+For more information, examples, and documentation, please visit our website:
+https://scikit-image.org
 
-def find_author_info(commit):
-    """Return committer and author of a commit.
+"""
+    label_section_map: tuple[str, str] = (
+        (":trophy: type: Highlight", "Highlights"),
```

Comment:
Should these be regexps?

Reply:
Mainly because the emojis, and people may edit those over time without thinking it affects anything. The patterns would be more robust to ignore those, e.g. ".*Highlight[s]?$".

Reply:
Done. I chose to make the patterns even more permissive for now.